python/pywatchman/aioclient.py

```diff
@@ -0,0 +1,286 @@
+        remove processing impacts both the unscoped and scoped stores
+        for the subscription data.
+        '''
+        # People may pass in Unicode strings here -- but currently BSER only
+        # returns bytestrings. Deal with that.
+        # TODO: This won't be a problem with BSER v2
+        if isinstance(root, str):
+            root = encoding.encode_local(root)
+            root = os.path.normcase(root)
+        if isinstance(name, str):
+            name = name.encode('utf-8')
+        self.__ensure_subscription_queue_exists(name, root)
+        return await self.sub_by_root[root][name].get()
+
+    async def pop_log(self):
+        '''Get one log from the log queue.'''
+        return await self.log_queue.get()
+
+    def close(self):
+        '''Close the underlying connection.'''
+        self.connection.close()
+
+    def __resolve_sockname(self):
+        # if invoked via a trigger, watchman will set this env var; we
+        # should use it unless explicitly set otherwise
+        path = os.getenv('WATCHMAN_SOCK')
+        if path:
+            return path
+
+        cmd = ['watchman', '--output-encoding=bser', 'get-sockname']
+        try:
+            p = subprocess.Popen(cmd,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.PIPE,
+                                 close_fds=os.name != 'nt')
+        except OSError as e:
+            raise WatchmanError('"watchman" executable not in PATH (%s)', e)
+
+        stdout, stderr = p.communicate()
+        exitcode = p.poll()
+
+        if exitcode:
+            raise WatchmanError('watchman exited with code %d' % exitcode)
+
+        result = bser.loads(stdout)
+        self.__check_error(result)
+        return result['sockname']
+
+    def __check_error(self, res):
+        if 'error' in res:
+            raise CommandError(res['error'])
+
+    def __is_unilateral(self, res):
+        return 'unilateral' in res and res['unilateral'] or \
+               'subscription' in res or \
+               'log' in res
+
+    def __ensure_subscription_queue_exists(self, name, root):
+        # Note this function must be called from an async function on only one
+        # event loop.
+        if not root in self.sub_by_root:
+            self.sub_by_root[root] = {}
+        if not name in self.sub_by_root[root]:
+            self.sub_by_root[root][name] = asyncio.Queue()
+
+    async def __process_unilateral_response(self, response):
+        if 'log' in response:
+            await self.logs.put(response['log'])
+
+        elif 'subscription' in response:
+            sub = response['subscription']
+            root = os.path.normcase(response['root'])
+            self.__ensure_subscription_queue_exists(sub, root)
+            await self.sub_by_root[root][sub].put(response)
+
+        elif self.__is_unilateral(response):
+            raise Exception('Unknown unilateral response.')
```

Comment:
Prefer creating a named exception class for your exceptions. That way your clients can catch them specifically and don't need to resort to Pokemon exception handling.